service/storage_proxy.cc

```diff
@@ -1348,12 +1348,13 @@ endpoints_to_replica_ids(const locator::token_metadata& tm, const inet_address_v
 }
 
 query::max_result_size storage_proxy::get_max_result_size(const query::partition_slice& slice) const {
-    // Unpaged and reverse queries.
-    // FIXME: some reversed queries are no longer unlimited.
-    // To filter these out we need to know if it's a single partition query.
-    // Take the partition key range as a parameter? This would require changing many call sites
-    // where the partition key range is not immediately available.
-    // For now we ignore the issue, return the 'wrong' limit for some reversed queries, and wait until all reversed queries are fixed.
+    if (_features.cluster_supports_separate_page_size_and_safety_limit()) {
+        auto max_size = _db.local().get_unlimited_query_max_result_size();
+        return query::max_result_size(max_size.soft_limit, max_size.hard_limit, query::result_memory_limiter::maximum_result_size);
```

Comment:
I do see an advantage for having three limits. The page size tells us when to break a page. The soft limit tells us when to warn (can be larger than page size, we we warn on single-row pages > 5MB). The hard limit tells us when to throw.

What do we do now when a single-row page is larger than 1MB?

Reply:
When paging is enabled, the result memory accounter only looks at the page size, so you won't see any warnings during a paged reads (unless there is a reversing reader hiding somewhere in the reader stack).

Reply:
Then there isn't value in providing those three values, because we'll only look at either the limits, or the page size, but not both.

It also means it's possible to oom during a paged read since there are no limit checks (by having a large tombstone prefix) but we know how to fix that.

Independently we can oom by having a huge collection, but that's done before accounting and we can't fix it here.


Maybe we should switch to a new verb. If the rules are "if this is a paged query, treat the soft limit as a page size and don't warn/fail, but if it's an unpaged query, warn on the soft limit and fail on the hard limit" (are they?) then they are too complicated.

Having those three separate values *without those rules* is valuable, but we can't change the rules in-flight.

Reply:
> Maybe we should switch to a new verb. If the rules are "if this is a paged query, treat the soft limit as a page size and don't warn/fail, but if it's an unpaged query, warn on the soft limit and fail on the hard limit" (are they?) then they are too complicated.

Yes, these are the rules currently.

> Having those three separate values without those rules is valuable, but we can't change the rules in-flight.

We can with a feature and some care, as done in this PR.